### ISSUE TITLE
Repo: In .vscode/settings.json, use eslint.rules.customizations to downgrade complaints

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#5548]
### Suggestion

I've started including this in all my repos' `.vscode/settings.json` files:

```json
"eslint.rules.customizations": [{ "rule": "*", "severity": "warn" }],
```

Reasoning: it's confusing and inconvenient for developers to have ESLint complaints displayed with the same red squigglies as TypeScript complaints.

Shall we?

**Comment by bradzacher** [comment]
> it's confusing and inconvenient for developers to have ESLint complaints displayed with the same red squigglies as TypeScript complaints.

Why is it confusing? I'm not sure I understand what's confusing about seeing errors raised by eslint highlighted the same as errors raised by typescript.

If we have things that are set as an error when they aren't actual problems, then we should downgrade them to warnings in the config instead to ensure we aren't diluting the signal.

**Comment by JoshuaKGoldberg** [comment]
For developers less familiar with the whole ESLint<>TypeScript<>TypeScript-ESLint setup, knowing which tool is causing a red squiggly can be a nontrivial task. They have to look at potentially multiple shown-in-the-same-tooltip editor hover boxes. Granted, those are much less common on this repo - but for them at least it's a helpful practice to disambiguate.

Personally, I find it irksome to have the red visually all over the editor when I'm refactoring code. Especially when there are `@typescript-eslint/no-unsafe-*` rules enabled. The red lint rules get in the way of seeing what's actually a TypeScript error during those refactors. There's no visual way to disambiguate them.

But, within the context of the _lint run_, they're meant to be errors, no?

**Comment by bradzacher** [comment]
> knowing which tool is causing a red squiggly can be a nontrivial task

I guess my point is - why does it matter which tool created the squiggly?
An error is an error, regardless of the source, IMO.

The way I always have advised people is that errors are things that should never be in the codebase, full stop.
Warnings are things that are permissible (likely via disables or by leaving the lint in if the codebase allows warnings).

So it doesn't matter if it's an eslint error or a TS error or any other tool - an error needs to be fixed.

If you downgrade them in the IDE then you create this disjoint in the DevX where your file looks fine enough (only warnings), but then your CI dies on it because those warnings are actually errors in disguise.

That's just been my experience trying to build a linting culture at meta. Things are done a little differently there in terms of linting - but the IDE matching the CI on PRs is something users notice and care about deeply. If there are inconsistencies then they notice and they get confused!

**Comment by JoshuaKGoldberg** [comment]
Fascinating: it almost sounds like we're both optimizing for _almost_ the same thing, reducing user confusion over what's an error!

> an error needs to be fixed

But, what is _error_ and what is _needs_? In the contexts of code quality and getting PRs merged, then yes, all lint reports are errors that need to be fixed. But in the context of code running or being unit tested, then many lint reports are indeed just warnings that your code might not be "good".

For example, I often start rules like this to get a feel for what they're going to be like:

```ts
type Options = any; // todo: not sure what I want to call this
type MessageIds = any; // todo: idk yet

export default util.createRule<Options, MessageIds>({ /* ... */ });
```

So, I think what I'm suggesting is making the idea of _"permissible"_ more flexible?

N.b. since we've had a couple back-and-forths on this, if I haven't convinced anybody on the thread then I'm ok with dropping the issue and keeping as-is. 😄

**Comment by bradzacher** [comment]
> But, what is error

An error is simply an error-level signal.

> and what is needs?

By needs I mean that you mustn't ever disable it or ignore it, and you must fix it.

At meta it's a little different because of the codebase size (>>>500k files) we cannot ESLint check the entire codebase in the CI for every PR.
So as the ESLint config has evolved over time (and as plugins / eslint core have been upgraded) lint errors and warnings get added into existing files.
So we were left with a decision - either enforce that all PRs have no lint messages on changed files, or allow lint messages in changed files.
In the spirit of "moving fast" and allowing people to focus on their work (rather than refactoring existing code), we went with the latter.

So our guidance to people (and our tooling enforces this) is that error-level lints must be fixed - you cannot commit a change that introduces an erorr-level lint. But warnings are permissible to be added (though are actively discouraged).

A good way to think of what I'm saying is this; imagine you don't have CI and you rely on an honour system amongst your team. How do you provide guidelines to your devs?
"Error-level signals **must** be fixed. Warning and lower signals are left up to you; but you *should* endeavour to fix it"

----

> In the contexts of code quality and getting PRs merged, then yes, all lint reports are errors that need to be fixed

Not necessarily! By default ESLint actually allows warnings!

> [Normally, if ESLint runs and finds no errors (only warnings), it will exit with a success exit status.](https://eslint.org/docs/latest/user-guide/command-line-interface#--max-warnings)

You can use the `--max-warnings` CLI flags to specify a threshold for the number of warnings to be allowed.

Which means that right now our CI doesn't actually enforce that there are no lints in a PR because the lint step will not error out if there are only warnings.
It still reports them and they'll be shown inline by github, but they won't fail the CI step.

> But in the context of code running or being unit tested, then many lint reports are indeed just warnings that your code might not be "good".

To play devil's advocate - one could say the same thing about TS types. If you type a function as accepting `number | undefined`, but knowingly never pass `undefined` - then the TS type error would just be a warning that your code might not be "good" even though it works fine!

----

Happy to keep chatting! This is something I spend a lot of time thinking about and working on so I'm happy to hear another POV!
